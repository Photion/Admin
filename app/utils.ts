

export const capitalize = (s: string) => {
  return s
    .replace(/[\W-_]+/g, '-')
    .split('-')
    .map(split => split[0].toUpperCase() + split.substring(1).toLowerCase())
    .join(' ');
};

export const toOption = (value: string) => ({ value, text: capitalize(value) });
