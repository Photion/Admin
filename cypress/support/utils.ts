

export const getFile = (path, mime): Promise<{ file: File; dataTransfer: DataTransfer }> => {
  return new Promise((resolve) => {
    cy.fixture(path, 'base64')
      .then(content => Cypress.Blob.base64StringToBlob(content, mime))
      .then((blob) => {
        const file = new File([blob], path, { type: mime });
        const dataTransfer = new DataTransfer();

        dataTransfer.items.add(file);

        resolve({ file, dataTransfer });
      });
  });
};
