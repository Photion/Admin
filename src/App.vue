<template>
  <Layout />
</template>

<script lang="ts">
import { defineComponent } from 'vue';

import Layout from '~/src/vue/layouts/Layout.vue';

export default defineComponent({
  components: {
    Layout,
  },
});
</script>
