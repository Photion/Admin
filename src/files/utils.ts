
export const getBuffer = async (file: File) => {
  const fr = new FileReader();

  const buffer: ArrayBuffer = await new Promise((resolve) => {
    fr.onload = () => resolve(fr.result as ArrayBuffer);
    fr.readAsArrayBuffer(file);
  });

  return buffer;
};

export const getText = async (file: File) => {
  const fr = new FileReader();

  const text: string = await new Promise((resolve) => {
    fr.onload = () => resolve(fr.result as string);
    fr.readAsDataURL(file);
  });

  return text;
};
