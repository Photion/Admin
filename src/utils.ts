
export interface SelectOption {
  value: string | number;
  text: string;
}

export const capitalize = (s: string) => {
  return s
    .replace(/[\W-_]+/g, '-')
    .split('-')
    .map(split => split[0].toUpperCase() + split.substring(1).toLowerCase())
    .join(' ');
};

export const toOption = (value: string): SelectOption => ({ value, text: capitalize(value) });
