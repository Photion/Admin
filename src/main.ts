import { createApp } from 'vue';

import '~/src/plugins/sentry';
import '~/src/assets/scss/main.scss';
import { appLoaded } from '~/src/state/service';
import { router } from '~/src/vue/router';
import FontAwesomeIcon from '~/src/plugins/font-awesome';
import App from '~/src/App.vue';

export const main = async () => {
  await createApp(App)
    .use(router)
    .component('FontAwesomeIcon', FontAwesomeIcon)
    .mount('#app');

  appLoaded.value = true;
};

main();
