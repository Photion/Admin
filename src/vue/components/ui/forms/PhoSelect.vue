<template>
  <div class="pho-field-wrapper relative">
    <label v-if="label" :for="identifier" class="pho-label">{{ label }}</label>
    <div class="relative">
      <button
        :id="`${identifier}:trigger`"
        :cy="`${reference}:trigger`"
        @click="fn.toggleMenu(!showOptions)"
        @focusend="fn.toggleMenu(false)"
        type="button"
        class="pho-field focus:outline-none"
        aria-haspopup="true"
        aria-expanded="true">
        <div class="flex">
          <div class="flex-grow text-left">{{ displayValue }}</div>
          <div style="margin-right: 5px">
            <FontAwesomeIcon v-if="showOptions" icon="times"  @click="fn.toggleMenu(false, $event)" :cy="`${reference}:trigger:close`"  />
            <FontAwesomeIcon v-else icon="chevron-down" :cy="`${reference}:trigger:open`" />
          </div>
        </div>
        <div v-if="showOptions" class="absolute left-0 w-full mt-2 w-56 rounded-md shadow-lg cursor-pointer" style="z-index: 10">
          <div class="rounded-md bg-white shadow-xs py" role="menu" aria-orientation="vertical" aria-labelledby="options-menu">
            <PhoSelectOption
              v-for="option in options"
              :key="option.value"
              :identifier="identifier"
              :option="option"
              :current="is.current(option)"
              :selected="is.selected(option)"
              @mouseenter="on.option.hover(option)"
              @option="on.option.click" />
          </div>
        </div>
      </button>
    </div>
    <input
      :id="identifier"
      :value="modelValue"
      :name="name"
      :cy="reference"
      class="hidden" />
  </div>
</template>

<script lang="ts">
import { defineComponent, ref, computed, PropType } from 'vue';
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome';

import { createIdentifier } from '~/src/vue/components/shared';
import PhoSelectOption from '~/src/vue/components/ui/forms/PhoSelectOption.vue';
import { SelectOption, SelectOptionValue } from '~/src/utils';

export default defineComponent({

  components: {
    PhoSelectOption,
    FontAwesomeIcon,
  },

  props: {
    cy: {
      type: String,
      default: () => '',
    },
    label: {
      type: String,
      default: () => '',
    },
    options: {
      type: Object as PropType<SelectOption[]>,
      default: () => [],
    },
    name: {
      type: String,
      default: () => '',
    },
    modelValue: {
      type: [Number, String, Object, Array] as PropType<SelectOptionValue | SelectOptionValue[]>,
      required: true,
    },
    multiple: {
      type: Boolean,
      default: () => false,
    },
  },

  setup(props, context) {
    const current = ref<SelectOption | null>(null);
    const showOptions = ref(false);

    const identifier = createIdentifier('field', props.name);
    const reference = props.cy || identifier;

    const displayValue = computed(() => {
      let target: SelectOptionValue[];

      if (Array.isArray(props.modelValue)) {
        target = props.modelValue;
      } else {
        target = [props.modelValue];
      }

      return props.options
        .filter((option) => target.includes(option.value))
        .map((option) => option.text)
        .join(', ');
    });

    /**
     * Toggles the options menu
     *
     * @param state
     * @param $event
     */
    const toggleMenu = (state: boolean, $event?: MouseEvent) => {
      if ($event) {
        $event.stopPropagation();
      }
      showOptions.value = state;
    };


    /**
     * Toggles an option
     *
     * @param option
     */
    const toggleOption = (option: SelectOption) => {
      if (Array.isArray(props.modelValue)) {
        const values = props.modelValue.slice();
        const index = values.indexOf(option.value);

        if (index === -1) {
          values.push(option.value);
        } else {
          values.splice(index, 1);
        }
        return values;
      } else {
        return option.value;
      }
    };

    /**
     * Returns whether an option is selected
     *
     * @param option
     */
    const isSelected = (option: SelectOption) => {
      if (Array.isArray(props.modelValue)) {
        return props.modelValue.includes(option.value);
      } else {
        return option.value === props.modelValue;
      }
    };

    /**
     * Handles an option selection/deselection event
     *
     * @param option
     */
    const onOption = (option: SelectOption) => {
      const value = toggleOption(option);
      context.emit('update:modelValue', value);

      if (typeof props.modelValue === 'string') {
        toggleMenu(false);
      }

    };

    return {
      showOptions,
      displayValue,
      identifier,
      reference,
      fn: {
        toggleMenu,
      },
      is: {
        current: (option: SelectOption) => option.value === current.value?.value,
        selected: isSelected,
      },
      on: {
        option: {
          hover: (option: SelectOption) => (current.value = option),
          click: onOption,
        },
      },
    };
  },

});
</script>
