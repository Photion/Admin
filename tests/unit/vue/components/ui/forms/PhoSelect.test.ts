import { mount } from '@vue/test-utils';

import { SelectOption } from '~/src/utils';
import PhoSelect from '~/src/vue/components/ui/forms/PhoSelect.vue';

interface ComponentProps {
  [key: string]: unknown;
  name?: string;
  modelValue?: string | string[];
  multiple?: boolean;
  options?: SelectOption[];
}

const getComponent = (props: ComponentProps = {}) => {
  return mount(PhoSelect, { props: {
    name: 'pho-select-comp',
    modelValue: 2,
    multiple: false,
    options: [
      { value: 1, text: 'First' },
      { value: 2, text: 'Second' },
      { value: 44, text: 'Fortyforth' },
    ],
    ...props,
  } });
};

describe('unit.vue.components.ui.forms.PhoSelect', () => {

  it('Mounts', () => {
    const wrapper = getComponent();
    expect(wrapper.element).toMatchSnapshot();
  });

  it('Displays the options menu', async () => {
    const wrapper = getComponent();
    await wrapper.get('[cy="field:pho-select-comp:trigger"]').trigger('click');

    expect(wrapper.element).toMatchSnapshot();
  });

  it('Selects a new option, changes its value', async () => {
    const wrapper = getComponent();
    await wrapper.get('[cy="field:pho-select-comp:trigger"]').trigger('click');
    await wrapper.get('[cy="field:pho-select-comp:option:44"]').trigger('click');

    const emitted = wrapper.emitted();

    expect(emitted).toMatchSnapshot();
  });

  it('Selects multiple', async () => {
    const wrapper = getComponent({ multiple: true, modelValue: [] });
    await wrapper.get('[cy="field:pho-select-comp:trigger"]').trigger('click');
    await wrapper.get('[cy="field:pho-select-comp:option:44"]').trigger('click');
    expect(wrapper.emitted()).toMatchSnapshot();

    // Since we don't have a parent component
    // the modelValue won't be updated by itself
    await wrapper.setProps({ modelValue: [44] });
    await wrapper.get('[cy="field:pho-select-comp:option:1"]').trigger('click');
    expect(wrapper.emitted()).toMatchSnapshot();
  });

  describe('Stores the current value in a hidden input', () => {
    const test = (wrapper: ReturnType<typeof getComponent>) => {
      expect((wrapper.get('[cy="field:pho-select-comp"]').element as HTMLInputElement).value).toMatchSnapshot();
    };

    it('Single', async () => {
      const wrapper = getComponent();
      test(wrapper);

      await wrapper.setProps({ modelValue: 44 });
      test(wrapper);
    });

    it('Multiple', async () => {
      const wrapper = getComponent({ multiple: true, modelValue: ['2'] });
      test(wrapper);

      await wrapper.setProps({ modelValue: ['2', '44'] });
      test(wrapper);
    });

  });


});
