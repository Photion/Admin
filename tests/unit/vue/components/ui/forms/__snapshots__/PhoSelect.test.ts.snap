// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit.vue.components.ui.forms.PhoSelect Displays the options menu 1`] = `
<div
  class="pho-field-wrapper relative"
>
  <!--v-if-->
  <div
    class="relative"
  >
    <button
      aria-expanded="true"
      aria-haspopup="true"
      class="pho-field"
      cy="field:pho-select-comp:trigger"
      id="field:pho-select-comp:trigger"
      type="button"
    >
      <div
        class="flex"
      >
        <div
          class="flex-grow text-left"
        >
          Second
        </div>
        <div
          style="margin-right: 5px;"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-times fa-w-11"
            cy="field:pho-select-comp:trigger:close"
            data-icon="times"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 352 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              class=""
              d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="absolute left-0 w-full mt-2 w-56 rounded-md shadow-lg cursor-pointer"
        style="z-index: 10;"
      >
        <div
          aria-labelledby="options-menu"
          aria-orientation="vertical"
          class="rounded-md bg-white shadow-xs py"
          role="menu"
        >
          
          <div
            class=""
            cy="field:pho-select-comp:option:1"
            id="field:pho-select-comp:option:1"
            role="menuitem"
            value="1"
          >
            <span>
              First
            </span>
            <!--v-if-->
          </div>
          <div
            class=""
            cy="field:pho-select-comp:option:2"
            id="field:pho-select-comp:option:2"
            role="menuitem"
            value="2"
          >
            <span>
              Second
            </span>
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-check-square fa-w-14 text-primary-600"
              data-icon="check-square"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class=""
                d="M400 480H48c-26.51 0-48-21.49-48-48V80c0-26.51 21.49-48 48-48h352c26.51 0 48 21.49 48 48v352c0 26.51-21.49 48-48 48zm-204.686-98.059l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.248-16.379-6.249-22.628 0L184 302.745l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.25 16.379 6.25 22.628.001z"
                fill="currentColor"
              />
            </svg>
          </div>
          <div
            class=""
            cy="field:pho-select-comp:option:44"
            id="field:pho-select-comp:option:44"
            role="menuitem"
            value="44"
          >
            <span>
              Fortyforth
            </span>
            <!--v-if-->
          </div>
          
        </div>
      </div>
    </button>
  </div>
  <input
    class="hidden"
    cy="field:pho-select-comp"
    id="field:pho-select-comp"
    name="pho-select-comp"
  />
</div>
`;

exports[`unit.vue.components.ui.forms.PhoSelect Mounts 1`] = `
<div
  class="pho-field-wrapper relative"
>
  <!--v-if-->
  <div
    class="relative"
  >
    <button
      aria-expanded="true"
      aria-haspopup="true"
      class="pho-field"
      cy="field:pho-select-comp:trigger"
      id="field:pho-select-comp:trigger"
      type="button"
    >
      <div
        class="flex"
      >
        <div
          class="flex-grow text-left"
        >
          Second
        </div>
        <div
          style="margin-right: 5px;"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-chevron-down fa-w-14"
            cy="field:pho-select-comp:trigger:open"
            data-icon="chevron-down"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              class=""
              d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <!--v-if-->
    </button>
  </div>
  <input
    class="hidden"
    cy="field:pho-select-comp"
    id="field:pho-select-comp"
    name="pho-select-comp"
  />
</div>
`;

exports[`unit.vue.components.ui.forms.PhoSelect Selects a new option, changes its value 1`] = `
Object {
  "update:modelValue": Array [
    Array [
      44,
    ],
  ],
}
`;

exports[`unit.vue.components.ui.forms.PhoSelect Selects multiple 1`] = `
Object {
  "update:modelValue": Array [
    Array [
      Array [
        44,
      ],
    ],
  ],
}
`;

exports[`unit.vue.components.ui.forms.PhoSelect Selects multiple 2`] = `
Object {
  "update:modelValue": Array [
    Array [
      Array [
        44,
      ],
    ],
    Array [
      Array [
        44,
        1,
      ],
    ],
  ],
}
`;

exports[`unit.vue.components.ui.forms.PhoSelect Stores the current value in a hidden input Multiple 1`] = `"2"`;

exports[`unit.vue.components.ui.forms.PhoSelect Stores the current value in a hidden input Multiple 2`] = `"2,44"`;

exports[`unit.vue.components.ui.forms.PhoSelect Stores the current value in a hidden input Single 1`] = `"2"`;

exports[`unit.vue.components.ui.forms.PhoSelect Stores the current value in a hidden input Single 2`] = `"44"`;
